Mono.Debugging/Mono.Debugging.Client/DebuggerSession.cs

```diff
@@ -1302,9 +1323,23 @@ internal protected void OnTargetDebug (int level, string category, string messag
 
 		internal protected void OnAssemblyLoaded (string assemblyLocation)
 		{
-			AssemblyLoaded?.Invoke (this, new AssemblyEventArgs (assemblyLocation));
+			var assembly = new Assembly (assemblyLocation);
+			lock (assemblies) {
```

Comment:
To avoid code duplication, you can replace these 5 lines by `OnAssemblyLoaded(assembly)`